source/engine/serialize/SceneSerialization.cpp

```diff
@@ -8,57 +8,175 @@
-
-    const auto entities = BuildFragmentEntityList(ecs.GetAll<Entity>(), entityFilter, ecs);
-    const auto entityMap = BuildEntityToFragmentIdMap(entities);
-    const auto entityInfos = BuildFragmentEntityInfos(entities, ecs, entityMap);
-    const auto header = SceneFragmentHeader{};
-    serialize::Serialize(stream, header);
-    serialize::Serialize(stream, assets);
-    serialize::Serialize(stream, entityInfos);
+    constexpr auto header = nc::SceneFragmentHeader{};
+    nc::serialize::Serialize(stream, header);
 }
 
-void LoadSceneFragment(std::istream& stream,
-                       ecs::Ecs ecs,
-                       asset::NcAsset& assetModule)
+void LoadHeader(std::istream& stream)
 {
-    auto header = SceneFragmentHeader{};
-    serialize::Deserialize(stream, header);
-    if (header.magicNumber != g_sceneFragmentMagicNumber)
+    auto header = nc::SceneFragmentHeader{};
+    nc::serialize::Deserialize(stream, header);
+    if (header.magicNumber != nc::g_sceneFragmentMagicNumber)
     {
-        throw NcError("Unexpected SceneFragment header");
+        throw nc::NcError("Unexpected SceneFragment header");
     }
 
-    if (header.version != g_currentSceneFragmentVersion)
+    if (header.version != nc::g_currentSceneFragmentVersion)
     {
-        throw NcError("Unexpected SceneFragment version");
+        throw nc::NcError("Unexpected SceneFragment version");
     }
+}
 
+void SaveAssets(std::ostream& stream, const nc::asset::AssetMap& assets)
+{
+    nc::serialize::Serialize(stream, assets);
+}
+
+void LoadAssets(std::istream& stream, nc::asset::NcAsset& assetModule)
+{
+    auto assets = nc::asset::AssetMap{};
+    nc::serialize::Deserialize(stream, assets);
+    assetModule.LoadAssets(assets);
+}
+
+void SaveEntities(std::ostream& stream,
+                  std::function<bool(nc::Entity)>& entityFilter,
+                  nc::SerializationContext& ctx)
+{
+    const auto entities = nc::BuildFragmentEntityList(ctx.ecs.GetAll<nc::Entity>(), entityFilter, ctx.ecs);
+    ctx.entityMap = nc::BuildEntityToFragmentIdMap(entities);
+    const auto entityInfos = nc::BuildFragmentEntityInfos(entities, ctx.ecs, ctx.entityMap);
+    nc::serialize::Serialize(stream, entityInfos);
+}
+
+void LoadEntities(std::istream& stream, nc::DeserializationContext& ctx)
+{
+    auto entities = std::vector<nc::FragmentEntityInfo>{};
+    nc::serialize::Deserialize(stream, entities);
+    std::ranges::for_each(entities, [&ctx](auto& entityData)
     {
-        auto assets = asset::AssetMap{};
-        serialize::Deserialize(stream, assets);
-        assetModule.LoadAssets(assets);
+        nc::RemapEntity(entityData.info.parent, ctx.entityMap);
+        ctx.entityMap.emplace(entityData.fragmentId, ctx.ecs.Emplace<nc::Entity>(entityData.info));
+    });
+}
+
+auto GetSerializablePools(nc::ecs::Ecs ecs)
+{
+    auto filter = std::views::filter(ecs.GetComponentPools(), [](auto pool){ return pool->HasSerialize(); });
```

Comment:
Hooray for views! This is neat

Reply:
I'm still waiting for `std::ranges::to` to get implemented as the cherry on top:
```cpp
return std::views::filter(...) | std::ranges::to<std::vector>();
```